graphql-server/pkg/handlers/auth.go

```diff
@@ -24,3 +28,76 @@ func NonceHandler(sessionConfig config.SessionConfig) gin.HandlerFunc {
 		c.String(200, nonce)
 	}
 }
+
+//nolint:errcheck
+func VerificationHandler(sessionConfig config.SessionConfig) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		signedNonce, err := c.Cookie("nonce")
+		if err != nil {
+			c.AbortWithError(401, fmt.Errorf("nonce is either missing or invalid: %s", err))
+			return
+		}
+
+		var body models.AuthenticationRequestData
+		if err := c.BindJSON(&body); err != nil {
+			c.AbortWithError(403, fmt.Errorf("data or signature is empty or invalid: %s", err))
+			return
+		}
+
+		data, err := base64.StdEncoding.DecodeString(body.SignDoc.Messages[0].Value.Data)
+		if err != nil {
+			c.AbortWithError(403, fmt.Errorf("unable to decode data: %s", err))
+			return
+		}
+
+		// Parse request body with ABNF
+		authenticationMsg := new(models.AuthenticationMessage)
+		if err := authenticationMsg.ParseAbnf(string(data)); err != nil {
+			c.AbortWithError(403, fmt.Errorf("failed to parse authentication message: %s", err))
+			return
+		}
+
+		// Verify nonce signature
+		if !hmacsha256util.Verify(sessionConfig.SignatureSecret, authenticationMsg.Nonce, signedNonce) {
+			c.AbortWithError(401, fmt.Errorf("nonce is either missing or invalid"))
+			return
+		}
+
+		// Verify signature
+		pubKey, err := body.Signature.PubKey.ToSecp256k1()
+		if err != nil {
+			c.AbortWithError(403, fmt.Errorf("failed to parse pubkey: %s", err))
+			return
+		}
+
+		decodedSignature, err := base64.StdEncoding.DecodeString(body.Signature.Signature)
+		if err != nil {
+			c.AbortWithError(403, fmt.Errorf("failed to verify signature: %s", err))
+			return
+		}
+
+		marshalledMessage, err := json.Marshal(body.SignDoc)
+		if err != nil {
+			c.AbortWithError(403, fmt.Errorf("failed to verify signature: %s", err))
+			return
+		}
+
+		sigValid := pubKey.VerifySignature(marshalledMessage, decodedSignature)
+
+		if !sigValid {
+			c.AbortWithError(403, fmt.Errorf("signature invalid"))
+			return
+		}
+
+		// Sign address
+		address := body.SignDoc.Messages[0].Value.Signer
+		signedAddress := hmacsha256util.Sign(sessionConfig.SignatureSecret, address)
```

Comment:
we need to sign `address + expiry`? otherwise the cookie expiry can be easily spoofed?

and we will want to have a plaintext cookie `auth_session` + signed `auth_session.sig`?

Reply:
realized that if we want to persist the session through other chains as well we will have to sign the address hex instead

Reply:
i think per chain is what we want? we have different secrets for each deployment anyways?

alternatively we can parse bech32 at validation logic, i think showing bech32 in the challenge message (which is user visible) is important